sanic/helpers.py

```diff
@@ -153,3 +153,25 @@ def import_string(module_name, package=None):
     if ismodule(obj):
         return obj
     return obj()
+
+
+class ContextObject:
+    """Storage of arbitrary values using attribute syntax.
```

Comment:
so I really do think we should use a dictionary instead of a `ContextObject`, this object quacks like a dict so it should just be a dict. A couple problems I can imagine:
1. there is no way to tell if an attribute is set explicitly as None or your just getting the default (while with a dict the `[key]` vs `.get(key)` syntax lets you choose when getting the default is the right choice)
2. this cleans up direct access `request.custom_context.user` but it makes variable access opaque `request.custom_context.__dict__[my_key]` or `getattr(request.custom_context, my_key)`
3. extending a dict to [addict](https://github.com/mewwts/addict) style access is easy and feels natural, where as going back to a dict from this feels like your missing and important decision the framework developers made
```
from addict import Dict

@app.middleware("request")
def extend_custom_context(request):
    request.custom_context = Dict(request.custom_context)
```

Reply:
> so I really do think we should use a dictionary instead of a `ContextObject`, this object quacks like a dict so it should just be a dict. A couple problems I can imagine:

We seem to consider it from very different angles. For you it is fundamentally a dict and thus you consider iteration an issue. For me it is really equivalent to adding member variables to `request`, but namespaced to `custom_context` so that extensions do not collide with Sanic core (in particular, so that we don't need to consider someone already using a variable name whenever we add a new member to `request`).

In the context of variables, the `.name` syntax is correct, and there is no need for iteration (just like you wouldn't iterate all members of the `request` object).

Let's consider a session store, using a request middleware to get session variables from a session cookie, and a response middleware for updating the cookie with new values from custom_context.

The session variables cannot be loaded directly to custom_context because then any other middlewares' variables would also get stored to session -- this is not what we want -- and this is also why I think that custom_context should not be iterated.

If custom_context was a dict, the middleware would have to add its storage as `request.custom_context["session"]` and then the handler would be using `request.custom_context["session"].get("foo")` to get a value from session. Even PHP with its `$_SESSION["foo"]` is not nearly as cumbersome.

As a matter of fact, I am proposing another name change to simply `ctx`. While `custom_context` is explanatory, it really is only a namespace and a very tedious name for such a thing. `request.ctx.session.get("foo")` seems far more reasonable.

> 1. there is no way to tell if an attribute is set explicitly as None or your just getting the default (while with a dict the `[key]` vs `.get(key)` syntax lets you choose when getting the default is the right choice)

You can find out by `"foo" in request.custom_context` (`True` for existing keys, even if their value is `None`), although I would expect most request handlers to be more interested on the truthiness of values than specifically which of *undefined*, `None` or even otherwise *falsy* they are.

Reply:
It turns out that such a session store already exists https://github.com/xen/sanic_session (although storing in memory instead of cookie).

Reply:
Thumb emoji vote: thumb-up for `ctx`, thumb-down for `custom_context`.

Please read the above messages first. If you think that some other very short name would be better, thumb-up and comment which.

Reply:
My intuition here is that `.name` syntax to me implies some sort of predefined schema / api, the whole point of this `custom_context` is we don't know ahead of time how people are going to use this . I think most people will instinctively have different reactions when they get an `AttributeError` and a `KeyError`, the first makes you think "oh I'm using this object wrong" the second makes you think "why didn't that key get set". Because no one really owns the custom context (extensions can add whatever items they want in any order) I think a dict better reflects the reality that these are not predefined `.name` attributes but instead attached at runtime. I get that `request.custom_context["session"]` is strictly 3 more characters than it needs to be but I think there is an intention that is conveyed there that is worth keeping.

I agree that iterating seems like the wrong behavior in most cases! Its super fragile with pluggable extensions. However maybe were both starting with the wrong user story! If you look at the session store that you link to its whole purpose is to create this `"session"` variable that really does feel like it should be a available as a `request.session` attribute. Its kind of an extension of the `Request` class api, and after I `Session(app, interface=InMemorySessionInterface())` I always expect `request.session` to be there just like I expect `request.content_type` to be there and I expect it to be defined in exactly one place. As is using a `dict` or a `CustomContext` both enable a class of bugs where multiple extensions may not know about each other and both try to set their own `"session"` i.e. `request.custom_context['session'] = foo` or `request.custom_context.session = bar`. I think maybe the real answer is to come up with a registry of request attributes where extensions can add their custom logic. i.e. something like:

```
# haven't thought about the signature / names too hard, open to suggestions
@app.request_ctx
def session(request: sanic.request.Request) -> sanic_session.SessionDict:
    # sanic_session logic goes here
    return ...

@app.route("/")
def index(request):
    my_session = request.session
    return ...
```

I think this has a lot of benefits
1. the api for extending sanic looks just like the api for using it
2. we can prevent name clashes at import time
2. we can implement a lot of the exiting request class using this kind of logic and dogfood our api
3. the request_ctx handler can be lazy but cache the result, this makes the order of execution of middle ware unimportant + mimics how we operate in the actual request class preventing unnecessary work.

tldr:
I think with how sanic currently works a dictionary just reflects the reality of the `custom_context` better however I think if we rework how extensions are implemented we end up with a better product and something that makes sense using that sweet sweet syntactical sugar of  `request.name`

Reply:
just realized @ashleysommer may have some thoughts on this as the author of https://github.com/ashleysommer/sanicpluginsframework